lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -2134,6 +2134,78 @@ static void optimizeQuantization(Function *F) {
   optimizeQuantizedMaxSplat(F);
 }
 
+#define BOILER_PLATE_TEMPLATE(NODE_NAME_)                                      \
+  template <typename Q = T, typename... ArgsTypes>                             \
+  static boilerPlateResolve<NODE_NAME_##Node, Q> *
+
+template <typename T> struct NodeCreator {
+  template <typename CmpType, typename InstType>
+  using boilerPlateResolve =
+      typename std::enable_if<std::is_same<InstType, CmpType>::value,
+                              InstType>::type;
+  BOILER_PLATE_TEMPLATE(AvgPool)
+  create(Function &F, ArgsTypes... var2) { return F.createAvgPool(var2...); }
+  BOILER_PLATE_TEMPLATE(MaxPool)
+  create(Function &F, ArgsTypes... var2) { return F.createMaxPool(var2...); }
+  BOILER_PLATE_TEMPLATE(Add)
+  create(Function &F, ArgsTypes... var2) { return F.createAdd(var2...); }
+  BOILER_PLATE_TEMPLATE(Sub)
+  create(Function &F, ArgsTypes... var2) { return F.createSub(var2...); }
+  BOILER_PLATE_TEMPLATE(Mul)
+  create(Function &F, ArgsTypes... var2) { return F.createMul(var2...); }
+  BOILER_PLATE_TEMPLATE(Div)
+  create(Function &F, ArgsTypes... var2) { return F.createDiv(var2...); }
+  BOILER_PLATE_TEMPLATE(Min)
+  create(Function &F, ArgsTypes... var2) { return F.createMin(var2...); }
+  BOILER_PLATE_TEMPLATE(Max)
+  create(Function &F, ArgsTypes... var2) { return F.createMax(var2...); }
+};
```

Comment:
I don't think the NodeCreator class is necessary.  I'm pretty sure it will work to have a set of free functions:
```
template<typename T, typename ...Args>
enable_if_same_t<T, AvgPoolNode> *create(Function &F, Args... args) {
  return F.createAvgPool(args...);
}
```

Reply:
@bertmaher Something like this?
enable_if_t is c++14, but I guess most people should be using compilers that support it by now.

```
template <class T, class U>
using enable_if_same_t = std::enable_if_t<std::is_same<T, U>::value, U>;

#define FUNCTION_ENABLE_IF_TEMPLATE(NODE_NAME_)                                \
  template <class T, typename... Args> enable_if_same_t<T, NODE_NAME_##Node>

FUNCTION_ENABLE_IF_TEMPLATE(AvgPool) * createNode(Function &F, Args... args) {
  return F.createAvgPool(args...);
}
```

Reply:
I left FUNCTION_ENABLE_IF_TEMPLATE because I think it kind of hides away some of the template stuff, and if anyone cares it's just couple of lines above to look at. That being said don't have strong feelings about it. Can replace it with explicit template stuff. Just don't like repetitive code like that.